### ISSUE TITLE
Verify the effect of non-ascii characters on paths to databases

### ISSUE THREAD

**Issue by borrrden** [couchbase/couchbase-lite-core#23]
This is especially risky on Windows which does not use UTF-8 at the OS level and will incorrectly interpret UTF-8 bytes in the system [code page](https://en.wikipedia.org/wiki/Code_page) (FYI the code page on my work computer is [932](https://en.wikipedia.org/wiki/Code_page_943)) resulting in an invalid path.  We will probably need to interface away the file access or use something like [boost nowide](http://cppcms.com/files/nowide/html/) or [boost filesystem](http://www.boost.org/doc/libs/1_62_0/libs/filesystem/doc/index.htm).

**Comment by snej** [comment]
We could add a platform-specific method to the Path class to convert a UTF-8 string to the filesystem's native encoding.

Cocoa has a method like this, `NSString.fileSystemRepresentation`. This basically encodes the NSString as UTF-8, but there are some nuances of the way it decomposes characters, to match the exact encoding the filesystem expects.

**Comment by borrrden** [comment]
Actually now that I think about it this will not cover our bases enough.  Code pages are ancient and cannot represent even close to the number of characters needed for unicode support.  So even if we convert like this we are left with Windows' solution is to use the wide char equivalents (i.e. utf-16) of its OS API to get unicode support:

Handy background and reference : http://utf8everywhere.org/

**Comment by borrrden** [comment]
The effects are this:

LiteCore: Doesn't handle UTF-8 strings, and the resulting strings when used in file I/O have their non-ASCII characters displayed in [Mojibake](https://en.wikipedia.org/wiki/Mojibake) in Explorer.  Techncially, as long as no illegal characters end up being written, this will be survivable if strange and annoying for people except for the fact that:

SQLite:  Handles UTF-8 strings, and when it looks in the correct place it will return an error because the file exists in the incorrectly encoded path instead of the correct one.

ForestDB: Doesn't handle UTF-8 strings, and so cannot sanely co-exist with SQLite behind the same interface.  In order to get back to what it expects you would have to convert to UTF-16, and then back to ACP (Active Code Page) since there is no direct conversion between multibyte character sets in Windows (One could be written but is that a good use of time?)

I was able to push up changes to get LiteCore working with UTF-8 strings on Windows to the `fix_utf8_windows` branch and I was able to fix *most* of the ForestDB errors by replacing just a couple of lines, but this would be a breaking change for them in terms of functionality so they might not be open to it (hopefully they will allow a compilation flag).  There is still one error deleting the old file as a result of compaction that I curiously have not been able to track down yet.  I filed an issue in JIRA about UTF-8 support.
